src/Algorithms/CARTTrainer.cpp

```diff
@@ -245,42 +244,43 @@ void CARTTrainer::measureStrength(TreeType & tree, std::size_t nodeId, std::size
 }
 
 //Classification case
-CARTTrainer::TreeType CARTTrainer::buildTree(AttributeTables const& tables, ClassificationDataset const& dataset, boost::unordered_map<std::size_t, std::size_t>& cAbove, std::size_t nodeId ){
-	//Construct tree
-	ModelType::NodeInfo nodeInfo;
-	nodeInfo.nodeId = nodeId;
-	nodeInfo.leftNodeId = 0;
-	nodeInfo.rightNodeId = 0;
-	// calculate the label of the node, which is the propability of class c 
+CARTTrainer::TreeType CARTTrainer::
+buildTree(AttributeTables const &tables,
+		  ClassificationDataset const &dataset,
+		  ClassVector &cAbove, std::size_t nodeId) {
+    //Construct tree, and
+	// calculate the label of the node, which is the propability of class c
 	// given all points in this split for every class
-	nodeInfo.label = hist(cAbove);
+	ModelType::NodeInfo nodeInfo{nodeId,hist(cAbove)};
+
 	// calculate the misclassification propability,
 	// 1-p(j*|t) where j* is the class the node t is most likely to belong to;
 	nodeInfo.misclassProp = 1- *std::max_element(nodeInfo.label.begin(), nodeInfo.label.end());
-	
+
 	//calculate leaves from the data
-	
+
 	//n = Total number of cases in the split
 	std::size_t n = tables[0].size();
 
 	if(!(gini(cAbove,n)==0 || n <= m_nodeSize)){
 		//Count matrices
-		
+
 
 		//search the split with the best impurity
 		double bestImpurity = n+1.0;
-		std::size_t bestAttributeIndex, bestAttributeValIndex;//index of the best split
-		boost::unordered_map<std::size_t, std::size_t> cBestBelow, cBestAbove;//labels of best split
+		std::size_t bestAttributeIndex = 0, bestAttributeValIndex = 0;//index of the best split
+		ClassVector cBelow(m_labelCardinality),cBestBelow(m_labelCardinality),
+				cBestAbove(m_labelCardinality);//labels of best split
 
-		for (std::size_t attributeIndex=0; attributeIndex < m_inputDimension; attributeIndex++){
+		for (std::size_t attributeIndex=0; attributeIndex < m_inputDimension; ++attributeIndex){
 			AttributeTable const& table = tables[attributeIndex];
-			boost::unordered_map<std::size_t, std::size_t> cTmpAbove = cAbove;
-			boost::unordered_map<std::size_t, std::size_t> cBelow;
-			for(std::size_t i=0; i<n-1; i++){//go through all possible splits
+			auto cTmpAbove = cAbove;
+			std::fill(cBelow.begin(),cBelow.end(),0);
```

Comment:
Unfortunately `std::fill` can't be replaced by `cBelow.clear`, because `clear` leaves the vector empty, when we need it to still have its length, just zeroed.

That said, the changes to CARTTrainer.* in this changeset only include the unordered_map → vector conversion. The next pull-request will include many of the same changes as done to the RFTrainers. After that the splitting will occur in much the same way as in RF, and we'll be done with `std::fill` :)